src/libs/sys-util-rs/src/mount.rs

```diff
@@ -0,0 +1,1052 @@
+        // Lazy unmounting the mountpoint with the MNT_DETACH flag.
+        umount2(path, true).map_err(|e| Error::Umount(path.to_owned(), e))?;
+        info!(sl!(), "lazy umount for {}", path.display());
+    } else {
+        let start_time = Instant::now();
+        while let Err(e) = umount2(path, false) {
+            match e.kind() {
+                // The mountpoint has been concurrently unmounted by other threads.
+                io::ErrorKind::InvalidInput => break,
+                io::ErrorKind::WouldBlock => {
+                    let time_now = Instant::now();
+                    if time_now.duration_since(start_time).as_millis() > timeout as u128 {
+                        warn!(sl!(),
+                                  "failed to umount {} in {} ms because of EBUSY, try again with lazy umount",
+                                  path.display(),
+                                  Instant::now().duration_since(start_time).as_millis());
+                        return umount2(path, true).map_err(|e| Error::Umount(path.to_owned(), e));
+                    }
+                }
+                _ => return Err(Error::Umount(path.to_owned(), e)),
+            }
+        }
+
+        info!(
+            sl!(),
+            "umount {} in {} ms",
+            path.display(),
+            Instant::now().duration_since(start_time).as_millis()
+        );
+    }
+
+    Ok(())
+}
+
+/// Umount all filesystems mounted at the `mountpoint`.
+///
+/// If `mountpoint` is empty or doesn't exist, `umount_all()` is a noop. Otherwise it will try to
+/// unmount all filesystems mounted at `mountpoint` repeatedly. For example:
+/// - bind mount /dev/sda to /tmp/mnt
+/// - bind mount /tmp/b to /tmp/mnt
+/// - umount_all("tmp/mnt") will umount both /tmp/b and /dev/sda
+///
+/// # Safety
+/// Caller needs to ensure safety of the `path` to avoid possible file path based attacks.
+pub fn umount_all<P: AsRef<Path>>(mountpoint: P, lazy_umount: bool) -> Result<()> {
+    if mountpoint.as_ref().is_empty() || !mountpoint.as_ref().exists() {
+        return Ok(());
+    }
+
+    loop {
+        match umount2(mountpoint.as_ref(), lazy_umount) {
+            Err(e) => {
+                // EINVAL is returned if the target is not a mount point, indicating that we are
+                // done. It can also indicate a few other things (such as invalid flags) which we
+                // unfortunately end up squelching here too.
+                if e.kind() == io::ErrorKind::InvalidInput {
+                    break;
+                } else {
+                    return Err(Error::Umount(mountpoint.as_ref().to_path_buf(), e));
+                }
+            }
+            Ok(()) => (),
+        }
+    }
+
+    Ok(())
+}
+
+// Counterpart of nix::umount2, with support of `UMOUNT_FOLLOW`.
+fn umount2<P: AsRef<Path>>(path: P, lazy_umount: bool) -> std::io::Result<()> {
+    let path_ptr = path.as_ref().as_os_str().as_bytes().as_ptr() as *const i8;
```

Comment:
`i8` will not work for all architectures, can we make that a [`std::os::raw::c_char`](https://doc.rust-lang.org/std/os/raw/type.c_char.html)?

Reply:
Good catch @Jakob-Naucke !

Reply:
fixed